onadata/apps/api/viewsets/xform_viewset.py

```diff
@@ -749,6 +763,74 @@ def destroy(self, request, *args, **kwargs):
 
         return Response(status=status.HTTP_204_NO_CONTENT)
 
+    @action(methods=['GET'], detail=True)
+    def versions(self, request, *args, **kwargs):
+        xform = self.get_object()
+        versions_cache_key = f'{XFORM_DATA_VERSIONS}{xform.pk}'
+        version_id = kwargs.get('version_id')
+        job_uuid = request.query_params.get('job_uuid')
+
+        versions = cache.get(versions_cache_key, [])
+
+        # Retrieve a list of all available versions from the XForms'
+        # instances if the number of submissions are not more than the
+        # retrieval threshold
+        if not xform.num_of_submissions > getattr(
+            settings, "SUBMISSION_RETRIEVAL_THRESHOLD", 10000
+        ) and not versions:
+            versions = list(
+                xform.instances.filter(
+                    deleted_at__isnull=True
+                ).values('version').annotate(total=Count('version'))
```

Comment:
It does not seem like we use the version counts in any of the responses below, why do we need this here?